runtime/near-evm-runner/src/lib.rs

```diff
@@ -0,0 +1,589 @@
+#[macro_use]
+extern crate enum_primitive_derive;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use ethereum_types::{Address, H160, U256};
+use evm::CreateContractAddress;
+
+use near_runtime_fees::RuntimeFeesConfig;
+use near_vm_errors::{EvmError, FunctionCallError, VMError};
+use near_vm_logic::gas_counter::GasCounter;
+use near_vm_logic::types::{AccountId, Balance, Gas, ReturnData, StorageUsage};
+use near_vm_logic::{ActionCosts, External, VMConfig, VMLogicError, VMOutcome};
+
+use crate::evm_state::{EvmAccount, EvmState, StateStore};
+use crate::types::{
+    AddressArg, GetStorageAtArgs, Result, TransferArgs, ViewCallArgs, WithdrawArgs,
+};
+use crate::utils::{ecrecover_address, near_erc721_domain, prepare_meta_call_args};
+
+mod builtins;
+mod evm_state;
+mod interpreter;
+mod near_ext;
+pub mod types;
+pub mod utils;
+
+pub struct EvmContext<'a> {
+    ext: &'a mut dyn External,
+    account_id: AccountId,
+    signer_id: AccountId,
+    predecessor_id: AccountId,
+    current_amount: Balance,
+    attached_deposit: Balance,
+    storage_usage: StorageUsage,
+    pub logs: Vec<String>,
+    gas_counter: GasCounter,
+    fees_config: &'a RuntimeFeesConfig,
+    domain_separator: [u8; 32],
+}
+
+enum KeyPrefix {
+    Account = 0,
+    Contract = 1,
+}
+
+fn address_to_key(prefix: KeyPrefix, address: &H160) -> Vec<u8> {
+    let mut result = Vec::with_capacity(21);
+    result.push(prefix as u8);
+    result.extend_from_slice(&address.0);
+    result
+}
+
+impl<'a> EvmState for EvmContext<'a> {
+    fn code_at(&self, address: &H160) -> Result<Option<Vec<u8>>> {
+        self.ext
+            .storage_get(&address_to_key(KeyPrefix::Contract, address))
+            .map(|value| value.map(|x| x.deref().unwrap_or(vec![])))
+    }
+
+    fn set_code(&mut self, address: &H160, bytecode: &[u8]) -> Result<()> {
+        self.ext.storage_set(&address_to_key(KeyPrefix::Contract, address), bytecode)
+    }
+
+    fn get_account(&self, address: &Address) -> Result<Option<EvmAccount>> {
+        self.ext.storage_get(&address_to_key(KeyPrefix::Account, address)).map(|value| {
+            value.map(|x| {
+                EvmAccount::try_from_slice(&x.deref().expect("Failed to deref")).unwrap_or_default()
+            })
+        })
+    }
+
+    fn set_account(&mut self, address: &Address, account: &EvmAccount) -> Result<()> {
+        self.ext.storage_set(
+            &address_to_key(KeyPrefix::Account, address),
+            &account.try_to_vec().expect("Failed to serialize"),
+        )
+    }
+
+    fn _read_contract_storage(&self, key: [u8; 52]) -> Result<Option<[u8; 32]>> {
+        self.ext
+            .storage_get(&key)
+            .map(|value| value.map(|x| utils::vec_to_arr_32(x.deref().expect("Failed to deref"))))
```

Comment:
`vec_to_arr_32` expects value to be `32` bytes and panics. 
Probably should wrap into storage error